File: designPhoneDirectory.py
'''
Time Complexity :O(1) 
Space Complexity :O(n) ---> for queue 
Did this code successfully run on Leetcode : yes
Any problem you faced while coding this :no
'''

from collections import deque
class PhoneDirectory:

    def __init__(self, maxNumbers: int):
        
        # initlize a queue
        self.queue = deque([])

        # enque the elements to the queue
        for i in range(0,maxNumbers):
            self.queue.append(i)
        
        # initlize a set
        self.allocated = set()
        
    def get(self) -> int:
        
        # chk if my queue is not empty
        if len(self.queue) != 0:
            # deque the element from the queue
            ele = self.queue.popleft()
            
            # add the element to the set
            self.allocated.add(ele)
            
            # return the ele
            return ele
        
        # return default
        return -1

    def check(self, number: int) -> bool:
        '''
        If the number is avaialble in the set, return False else return True
        '''        
        if number in self.allocated:
            return False
        return True
        
    def release(self, number: int) -> None:
        '''
        Remove the element from the set and allocate it back into the queue
        '''
        if number in self.allocated:
            # remove the number from the set
            self.allocated.remove(number)
        
            # enque the number back to the queue
            self.queue.append(number)


# Your PhoneDirectory object will be instantiated and called as such:
# obj = PhoneDirectory(maxNumbers)
# param_1 = obj.get()
# param_2 = obj.check(number)
# obj.release(number)

File: DesignPhoneDirectory.cpp
//Time Complexity O(n)
// Space Complexity O(n)
//Problem successfully executed on leetcode
//No problems faced while coading this.


#include<iostream>
#include<vector>
#include<queue>
#include<map>
#include<unordered_set>
#include<stack>
using namespace std;

queue<int> q;
unordered_set<int> set;
public PhoneDirectory(int maxNos)
    {
        for(int i=0;i<maxNos;i++)
        {
            set.insert(i);
            q.push(i);
        }
    }
    
    public int Get()
    {
        if(q.empty()) return -1;
        int ans=q.front();
        q.pop();
        set.erase(ans);
    }
    
    public bool check(int number)
    {
        if(set.find(number)!=set.end())
        {
            return true;
        }
        return false;
    }
    public void release(int number)
    {
        if(set.find(number)!=set.end())
            return;
        set.insert(number);
        q.push(i);
    }
